feat: Add StageMetadataChecker Validator for core USD

beersandrew opened the pull request:
### Description of Change(s)
- adding validation for defaultPrim
- including necessary updates to associated plugInfo.json & CMakeLists.txt files

### Fixes Issue(s)
- #3122

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [x] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [X] I have submitted a signed Contributor License Agreement

beersandrew committed:
feat: start implementing StageMetadataChecker

- adding entry into plugInfo.json
- adding tokens
- adding function itself

Note: this does not work for upAxis or metersPerUnit because I'm not sure how to include the reference to UsdGeom

beersandrew committed:
feat: finish initial implementation

- trimmed down the functionality to only check for a defaultPrim, since the other fields will go in usdGeom
- added a test file, included in CMakeLists.txt

jesschimein commented:
Filed as internal issue #USD-9770

jesschimein commented:
/AzurePipelines run

beersandrew committed:
refactor: update names, docs

- corrected doc to only include defaultPrim check

tallytalwar commented on pxr/usd/usd/coreValidators.cpp:
```diff
@@ -12,26 +12,44 @@
+    if (!usdStage->GetDefaultPrim()) {
+        errors.emplace_back(UsdValidationErrorType::Error,
+                            UsdValidationErrorSites{UsdValidationErrorSite(usdStage, SdfPath("/"))},
+                            "Stage has missing or invalid defaultPrim.");
```
Maybe include the path to the rootLayer here in the reported error string.

beersandrew replied:
Updated

tallytalwar commented on pxr/usd/usd/plugInfo.json:
```diff
@@ -148,7 +148,10 @@
-                    }, 
+                    },
+                    "StageMetadataChecker": {
+                        "doc": "Stages that can be consumed as referencable assets should have a valid 'defaultPrim' declared."
```
Its a error and not a warning, so maybe "must have a valid 'defaultPrim' specified."?

beersandrew replied:
Updated

tallytalwar commented on pxr/usd/usd/testenv/testCoreUsdValidators.cpp:
```diff
@@ -0,0 +1,48 @@
+void TestCoreUsdStageMetadata()
+{
+
+    // Get stageMetadataChecker
```
This is good, but I will also add a test of a valid stage which provides a defaultPrim.

beersandrew replied:
Updated

tallytalwar reviewed (commented):
Thanks @beersandrew 

This looks great, just a couple of nitpicks. I will get this synced internally within Pixar code as soon as these are addressed. Really appreciate the help here.

tallytalwar commented on pxr/usd/usd/CMakeLists.txt:
```diff
@@ -299,6 +299,13 @@ pxr_build_test_shared_lib(TestUsdResolverChangedResolver
         testenv/TestUsdResolverChangedResolver.cpp
 )
 
+pxr_build_test(testCoreUsdValidators
```
Also this makes me realize I should have included my CompositionErrorValidator test in something like this, instead of bundling it in the validationRegistry test. I will do move it internally. But thanks!

tallytalwar commented:
Also, I would also update this test: http://grok.pixar.com/xref/mainline/pxr/usd/usd/testenv/testUsdValidationRegistry.cpp?r=2331214#225-240 to include this StageMetadata test now.

I also realize this test should be a subset test like I did for usdShade: https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp#L22-L46 (May I suggest while you are updating the test to include StageMetadata validator, to also modify it to be a subset test like with usdShade). Please let me know if this doesn't make sense and we can chat about it. Thanks

beersandrew committed:
fix: clean up based on PR comments

- fixed description string in plugInfo.json to be clear about requiring a defaultPrim
- updated validation error message to include the rootLayer identifier
- added the defaultPrim at the end of the unit test to verify the error is cleared when a defaultPrim is set

beersandrew committed:
test: update and separate validator test

- add metadata test to validator check
- separate validator check into a new test

beersandrew commented:
> Also, I would also update this test: http://grok.pixar.com/xref/mainline/pxr/usd/usd/testenv/testUsdValidationRegistry.cpp?r=2331214#225-240 to include this StageMetadata test now.
> 
> I also realize this test should be a subset test like I did for usdShade: https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp#L22-L46 (May I suggest while you are updating the test to include StageMetadata validator, to also modify it to be a subset test like with usdShade). Please let me know if this doesn't make sense and we can chat about it. Thanks

I think this is what you meant? : https://github.com/PixarAnimationStudios/OpenUSD/pull/3125/commits/be2078039b0c3fa9f9f4a7b8e48759266029609a

I'm getting an error when I run the validator registry test locally before making that change:

`testUsdValidationRegistry crashed. FATAL ERROR: Failed axiom: ' !PlugRegistry::GetInstance().RegisterPlugins(testDir).empty() '`

Let me know if this is what you meant and if not I may need help getting my local setup to pass the `testUsdValidationRegistry` test to make sure my changes are working before committing

tallytalwar commented on pxr/usd/usd/testenv/testUsdValidationRegistry.cpp:
```diff
@@ -310,6 +294,27 @@ void TestUsdValidationRegistry()
+    TF_AXIOM(coreValidatorMetadata.size() == 1);
+    const std::vector<TfToken> expectedValidatorNames =
+            {UsdValidatorNameTokens->compositionErrorTest,
+             UsdValidatorNameTokens->stageMetadataChecker};
```
Thanks. But since you are touching this code, can I request you to update this to the subset check instead of equality check (Similar to this https://github.com/PixarAnimationStudios/OpenUSD/blob/dev/pxr/usd/usdShade/testenv/testUsdShadeValidators.cpp#L22-L46).

Reasoning being, if a client is running this test in their own environment and they have validators available in their plugins which are listing "UsdCoreValidators" as a keyword, the equality will fail, so all we can do here is to make sure the validator provides by openUSD under UsdCoreValidators keyword are found.

tallytalwar replied:
And you mentioned the following assert is failing for you?

`TF_AXIOM(!PlugRegistry::GetInstance().RegisterPlugins(testDir).empty());`

- This could be because the "pxr/usd/usd/testenv/testUsdValidationRegistry" dir is not being installed property for you .. I don't know why that will be though.
- Are you running the test via ctest? 
- The test uses [ArchGetCwd()](https://github.com/PixarAnimationStudios/OpenUSD/blob/be2078039b0c3fa9f9f4a7b8e48759266029609a/pxr/usd/usd/testenv/testUsdValidationRegistry.cpp#L322C33-L322C43), which should have returned the path where testUsdValidationRegistry is installed.

I checked all build configurations internally and this test does pass in all.

Please let me know if you still need help with this.

beersandrew replied:
updated to do the subset check rather than equality, sorry i missed that the first time around.

following up on slack about testing setups

tallytalwar reviewed (commented):
Provided notes on testUsdValidationRegistry inline. Let me know (here or on slack) if you are still running into test failure without your change. Thanks

jesschimein commented:
/AzurePipelines run

beersandrew committed:
test: update test

- allow validators to be contained in the set rather than requiring matching by name

tallytalwar commented on pxr/usd/usd/testenv/testCoreUsdValidators.cpp:
```diff
@@ -0,0 +1,59 @@
+
+    // Create an empty stage
+    SdfLayerRefPtr rootLayer = SdfLayer::CreateAnonymous();
+    UsdStageRefPtr usdStage = UsdStage::Open(rootLayer);
```
You could have directly done UsdStage::CreateInMemory here, and then you do not need to use the SdfLayer at all.

beersandrew replied:
I wanted to create a layer here to verify that the root layer string is in the error message (used lower on line 40). when I did it without a root layer, the error message prints "<>" which didn't feel like a great test. Happy to remove the layer if you'd like

tallytalwar replied:
I see, yeah makes sense. Sorry I missed the check on line 40.

tallytalwar reviewed (approved):
Just 1 tiny cleanup, and all great. Thanks for the patience and doing this work @beersandrew .

tallytalwar commented on pxr/usd/usd/testenv/testCoreUsdValidators.cpp:
Sorry noticed it when internalizing. (Taking care of it internally). But FYI

As a convention we prefix the test with the library name: so testUsdCoreValidators instead of testCoreUsdValidators.

jesschimein commented:
/AzurePipelines run
